lib/elixir/lib/kernel.ex

```diff
@@ -3029,16 +3029,23 @@ defmodule Kernel do
     in_module? = (__CALLER__.context == nil)
 
     right = case bootstrapped?(Macro) do
-      true  -> Macro.expand(right, __CALLER__)
+      true  -> Macro.prewalk(right, &Macro.expand(&1, __CALLER__))
```

Comment:
This change will also help fix the error message. If in_module? is false and the expansion does not return a list, then you can raise something meaningful.